driver/src/metrics/stablex_metrics.rs

```diff
@@ -0,0 +1,195 @@
+    ) {
+        let label = &["solution"];
+        self.processing_times
+            .with_label_values(label)
+            .observe(time_elapsed_since_batch_start(batch));
+        match res {
+            Ok(solution) => {
+                let touched_orders = orders
+                    .iter()
+                    .zip(&solution.executed_buy_amounts)
+                    .filter(|(_, amount)| amount > &&0)
+                    .map(|(o, _)| o.clone())
+                    .collect::<Vec<Order>>();
+                self.orders
+                    .with_label_values(label)
+                    .inc_by(touched_orders.len().try_into().unwrap_or(std::i64::MAX));
+                self.tokens
+                    .with_label_values(label)
+                    .inc_by(tokens_from_orders(&touched_orders));
+                self.users
+                    .with_label_values(label)
+                    .inc_by(users_from_orders(&touched_orders));
+            }
+            Err(_) => self.failures.with_label_values(label).inc(),
+        }
+    }
+
+    pub fn auction_solution_verified(&self, batch: U256, res: &Result<U256, DriverError>) {
+        let label = &["verification"];
+        self.processing_times
+            .with_label_values(label)
+            .observe(time_elapsed_since_batch_start(batch));
+        match res {
+            Ok(_) => (),
+            Err(_) => self.failures.with_label_values(label).inc(),
+        }
+    }
+
+    pub fn auction_solution_submitted(&self, batch: U256, res: &Result<(), DriverError>) {
+        let label = &["submission"];
+        self.processing_times
+            .with_label_values(label)
+            .observe(time_elapsed_since_batch_start(batch));
+        match res {
+            Ok(_) => self.successes.with_label_values(label).inc(),
+            Err(_) => self.failures.with_label_values(label).inc(),
+        }
+    }
+
+    pub fn auction_skipped(&self, batch: U256) {
+        let label = &["skipped"];
+        self.processing_times
+            .with_label_values(label)
+            .observe(time_elapsed_since_batch_start(batch));
+        self.successes.with_label_values(label).inc();
+    }
+}
+
+fn time_elapsed_since_batch_start(batch: U256) -> f64 {
+    let now = Utc::now().timestamp() as u64;
+    // A new batch is created every 5 minutes and becomes solvable one batch later
+    let batch_start = (batch.low_u64() + 1) * 300;
+    (now - batch_start) as f64
+}
+
+fn tokens_from_orders(orders: &Vec<Order>) -> i64 {
+    orders
+        .iter()
+        .flat_map(|order| vec![order.buy_token, order.sell_token].into_iter())
+        .collect::<HashSet<_>>()
+        .len()
+        .try_into()
+        .unwrap_or(std::i64::MAX)
+}
+
+fn users_from_orders(orders: &Vec<Order>) -> i64 {
+    orders
+        .iter()
+        .map(|order| order.account_id)
+        .collect::<HashSet<_>>()
+        .len()
+        .try_into()
+        .unwrap_or(std::i64::MAX)
+}
```

Comment:
I would think that these belong in the `models/orders.rs` and that tokens from orders can be used in the place where we construct the JSON instance to be passed to the solver.

Reply:
This calculates the number of tokens.
In `linear_optimization_price_finder.rs` it needs the full `HashSet` and turns it into a `Vec<String>`.
The function is short and both places aren't doing the same thing so I would keep it separate.